src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/Snapshots/ReferenceItem.cs

```diff
@@ -12,7 +12,7 @@ internal class ReferenceItem
     {
         // If additional fields/properties are added to this class, please update RestoreHasher
 
-        public ReferenceItem(string name, ImmutableArray<ReferenceProperty> properties)
+        public ReferenceItem(string name, IImmutableDictionary<string, string> properties)
```

Comment:
General comment on this PR: consider using the concrete type rather than the interface.

The concrete type has extension methods and behaviours that give better performance.

The interface gives more flexibility, but we're not using that currently and all these interfaces are internal as far as I can tell, so we could always widen to interfaces in future if needed.

Reply:
I think I used the interfaces only where I needed to.

Tracing this specific instance, it comes from `IProjectRuleSnapshot`, so CPS would need to be changed to return the concrete type in order to be able to use it here. Or you'd need to cast it and hope, or cast with a fallback that allocates. In any case, I'm going to consider this out of scope for my PR, and I'll let your team manage changes if you'd like to reduce interface usage.

Reply:
If it's coming from CPS then it's fine. CPS doesn't use the built in immutable collection types. It has its own optimized versions.